File: Client/src/app/features/auth/authorization/role-permissions.ts
import { APPLICATION_PERMISSIONS, ApplicationPermission } from './application-permission';

import { ApplicationRole} from '../data-access/models/application-role';

const ROLE_PERMISSIONS:
  Readonly<Record<ApplicationRole, readonly ApplicationPermission[]>> = {
    Unknown: [],

    ApplicationAdministrator: [
      APPLICATION_PERMISSIONS.dashboardReadAll,
      APPLICATION_PERMISSIONS.assetsReadAll,
      APPLICATION_PERMISSIONS.employeesReadAll,
      APPLICATION_PERMISSIONS.assetsManage,
      APPLICATION_PERMISSIONS.storesManage,
      APPLICATION_PERMISSIONS.usersManage
    ],

    CentralUser: [
      APPLICATION_PERMISSIONS.dashboardReadAll,
      APPLICATION_PERMISSIONS.assetsReadAll
    ],

    StoreManagement: [
      APPLICATION_PERMISSIONS.dashboardReadOwnStore,
      APPLICATION_PERMISSIONS.assetsReadOwnStore,
      APPLICATION_PERMISSIONS.employeesReadOwnStore
    ],

    StoreAdministrator: [
      APPLICATION_PERMISSIONS.dashboardReadOwnStore,
      APPLICATION_PERMISSIONS.assetsReadOwnStore,
    ]
  };

export function getRolePermissions(role: ApplicationRole): readonly ApplicationPermission[] {
  return ROLE_PERMISSIONS[role];
}

export function hasPermission(role: ApplicationRole, permission: ApplicationPermission): boolean {
  return ROLE_PERMISSIONS[role].includes(permission);
}

export function hasAnyPermission(role: ApplicationRole, permissions: readonly ApplicationPermission[]): boolean {
  return permissions.some(permission => hasPermission(role, permission)
  );
}

File: AssetManagement.Application/Common/Authorization/RolePermissions.cs
namespace AssetManagement.Application.Common.Authorization;

public static class RolePermissions
{
    private static readonly IReadOnlyDictionary<string, IReadOnlySet<string>>
        PermissionsByRole =
            new Dictionary<string, IReadOnlySet<string>>(
                StringComparer.Ordinal)
            {
                [ApplicationRoleNames.ApplicationAdministrator] =
                    new HashSet<string>(StringComparer.Ordinal)
                    {
                        ApplicationPermissions.DashboardReadAll,
                        ApplicationPermissions.AssetsReadAll,
                        ApplicationPermissions.EmployeesReadAll
                    },

                [ApplicationRoleNames.CentralUser] =
                    new HashSet<string>(StringComparer.Ordinal)
                    {
                        ApplicationPermissions.DashboardReadAll,
                        ApplicationPermissions.AssetsReadAll
                    },

                [ApplicationRoleNames.StoreManagement] =
                    new HashSet<string>(StringComparer.Ordinal)
                    {
                        ApplicationPermissions.DashboardReadOwnStore,
                        ApplicationPermissions.AssetsReadOwnStore,
                        ApplicationPermissions.EmployeesReadOwnStore
                    },

                [ApplicationRoleNames.StoreAdministrator] =
                    new HashSet<string>(StringComparer.Ordinal)
                    {
                        ApplicationPermissions.DashboardReadOwnStore,
                        ApplicationPermissions.AssetsReadOwnStore,
                        ApplicationPermissions.EmployeesReadOwnStore
                    }
            };

    public static IReadOnlySet<string> GetPermissions(string roleName)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(roleName);

        return PermissionsByRole.TryGetValue(
            roleName,
            out var permissions)
                ? permissions
                : new HashSet<string>(StringComparer.Ordinal);
    }

    public static bool HasPermission(
        string roleName,
        string permission)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(roleName);
        ArgumentException.ThrowIfNullOrWhiteSpace(permission);

        return GetPermissions(roleName).Contains(permission);
    }

    public static IReadOnlySet<string> GetPermissions(
        IEnumerable<string> roleNames)
    {
        ArgumentNullException.ThrowIfNull(roleNames);

        var permissions = new HashSet<string>(
            StringComparer.Ordinal);

        foreach (var roleName in roleNames)
        {
            if (string.IsNullOrWhiteSpace(roleName))
            {
                continue;
            }

            permissions.UnionWith(GetPermissions(roleName));
        }

        return permissions;
    }

    public static bool HasPermission(
        IEnumerable<string> roleNames,
        string permission)
    {
        ArgumentNullException.ThrowIfNull(roleNames);
        ArgumentException.ThrowIfNullOrWhiteSpace(permission);

        return roleNames.Any(
            roleName =>
                !string.IsNullOrWhiteSpace(roleName) &&
                HasPermission(roleName, permission));
    }
}